tests/reftests/shared-fetch.test

```diff
@@ -0,0 +1,240 @@
+  HASH=$(openssl $h $arch.tgz | cut -d' ' -f2)
+  echo "\"$h=$HASH\"" >> $file
+done
+echo "  ]" >> $file
+echo "}" >> $file
+### sh mkurl.sh extra.3 md5 sha512
+### sh mkurl.sh no-extra.3 sha256 sha512
+### sh mkurl.sh intra.3 sha512
+### sh mkurl.sh out.3 md5
+### opam update
+
+<><> Updating package repositories ><><><><><><><><><><><><><><><><><><><><><><>
+[default] synchronised from file://${BASEDIR}/REPO
+Now run 'opam upgrade' to apply any package updates.
+### opam remove no-extra
+The following actions will be performed:
+  - remove extra    3 [uses no-extra]
+  - remove no-extra 3
+===== 2 to remove =====
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> removed   extra.3
+-> removed   no-extra.3
+Done.
+### opam install out
+The following actions will be performed:
+  - install no-extra 3 [required by extra]
+  - install extra    3 [required by intra]
+  - install intra    3 [required by out]
+  - install out      3
+===== 4 to install =====
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> retrieved extra.3, intra.3, no-extra.3, out.3  (cached)
+-> installed no-extra.3
+-> installed extra.3
+-> installed intra.3
+-> installed out.3
+Done.
+### opam remove no-extra
+The following actions will be performed:
+  - remove out      3 [uses intra]
+  - remove intra    3 [uses extra]
+  - remove extra    3 [uses no-extra]
+  - remove no-extra 3
+===== 4 to remove =====
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> removed   out.3
+-> removed   intra.3
+-> removed   extra.3
+-> removed   no-extra.3
+Done.
+### <pkg:extra.3>
+opam-version: "2.0"
+build: [ "ocaml" "shared.ml" ]
+### <mkurl.sh>
+p=extra.3
+arch=shared
+file="REPO/packages/${p%.*}/$p/opam"
+echo "url {" >> $file
+echo "src: \"$arch.tgz\"" >> $file
+MD5="01234567890123456789012345678901"
+echo "checksum: \"md5=$MD5\"" >> $file
+echo "}" >> $file
+### sh mkurl.sh
+### opam update
+
+<><> Updating package repositories ><><><><><><><><><><><><><><><><><><><><><><>
+[default] synchronised from file://${BASEDIR}/REPO
+Now run 'opam upgrade' to apply any package updates.
+### opam install out | '[0-9a-f]{32,}' -> "a-given-hash"
+The following actions will be performed:
+  - install extra 3 [required by intra]
+  - install intra 3 [required by out]
+  - install out   3
+===== 3 to install =====
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+[ERROR] Conflicting file hashes, or broken or compromised cache!
+          - md5=a-given-hash (MISMATCH)
+          - md5=a-given-hash (match)
+          - sha512=a-given-hash (match)
+
+[ERROR] extra.3, intra.3, out.3: Checksum mismatch for file://${BASEDIR}/shared.tgz:
+          expected md5=a-given-hash
+          got      md5=a-given-hash
+[ERROR] Failed to get sources of extra.3, intra.3, out.3 (file://${BASEDIR}/shared.tgz): Checksum mismatch
+
+OpamSolution.Fetch_fail("Checksum mismatch")
```

Comment:
Can’t we catch this and fail gracefully instead? Also I think the three errors above are a bit redundant

Reply:
for that we need to fully rework error handling system...

Reply:
I agree with @kit-ty-kate here, but the situation is almost the same before this PR ([thanks to tests added before code :+1:](https://github.com/ocaml/opam/pull/4893/commits/ca040c65b126b9132e0c13fe48488d9964800e6b#diff-4ff8f50ddeb5f06fc5b5c42cf768bad7a2082bf230668ae0f2e8735081ae6ec9R236-R246))

Probably worth opening an issue. I could live with the redundant message — it's not something that you should see often after all — but the `OpamSolution.Fetch_fail` really is ugly.
IIRC a long time ago we tried to get rid of some of the redundant messages, but it's very difficult to ensure that you won't end up in cases where the message you removed was actually the only one appearing and as a result the error is not printed at all :shrug: